### PR TITLE
Release Candidate 0.3.2

### PR DESCRIPTION
**Pull request by jklenzing**
# Description

Release candidate for pysatMisisons.  Improves OS compatibility.

# Type of change

- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

This has been tested by running examples and unit tests on github actions.

Test pypi version instructions:
https://github.com/pysat/pysatMissions/wiki/v0.3.2---install-instructions-for-test-pypi

## Test Configuration
* Operating system: OS X Big Sur
* Version number: Python 3.8.11
* Any details about your local setup that are relevant: tested with pysat 3.0.4 and develop

# Checklist:

- [x] Update the version number
- [x] Verify and update Zenodo information in .zeondo.json
- [x] doc tests pass locally
- [x] activate this branch on readthedocs: https://pysatmissions.readthedocs.io/en/rc_0_3_2/
- [x] Upload the test distribution to the testPyPi server: https://test.pypi.org/project/pysatMissions/0.3.2rc2/
- [x] Contacted developers and users about the future release
- [x] Performed integration tests with downstream packages (include list of these packages) - na
- [x] Create a wiki page with the release text: https://github.com/pysat/pysatMissions/wiki/v0.3.2
- [ ] After approval, confirm all changes from the RC pull are listed in the wiki release text.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -63,17 +62,19 @@ python setup.py install
 
 Note: pre-1.0.0 version
 -----------------------
-pysatMissions is currently in an initial development phase and requires pysat 3.0.0.  
+pysatMissions is currently in an initial development phase and requires pysat
+3.0.2.
```
Comment (jklenzing) [inline README.md:66]:
```suggestion
3.0.4.  
```

File: README.md [hunk 2]
```diff
@@ -32,12 +31,12 @@ pysatMissions uses common Python modules, as well as modules developed by
...
-| numpy          | aacgmv2           |
-| pandas         | apexpy            |
-| pyEphem        | OMMBV>=1.0        |
-| sgp4>=2.7      | pysat>=3.0        |
+| Common modules | Community modules | Optional modules |
+| -------------- | ----------------- | ---------------- |
+| numpy          | pysat>=3.0.4      | aacgmv2          |
+| pandas         |                   | apexpy           |
+| pyEphem        |                   | OMMBV            |
+| sgp4>=2.7      |                   |                  |
```
Comment (aburrell) [inline README.md:39]: Are pyEphem and sgp4 also community modules?

File: pysatMissions/methods/spacecraft.py [hunk 3]
```diff
@@ -173,18 +163,85 @@ def project_ecef_vector_onto_sc(inst, x_label, y_label, z_label,
...
+
+    """
+
+    vector = np.vstack([x, y, z])
+    xhat, yhat, zhat = vector / np.linalg.norm(vector, axis=0)
+
+    return xhat, yhat, zhat
+
+
+def cross_product(x1, y1, z1, x2, y2, z2):
```
Comment (aburrell) [inline pysatMissions/methods/spacecraft.py:219]: How does this differ from `numpy.cross`?

File: pysatMissions/tests/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    def test_package_check_warns(self, package_name, num_warns):
+        """Test that package_check warns users if packages are not installed.
+
+        Parameters
+        ----------
+        package_name : str
+            Name of package to check for.
+        num_warns : int
+            Expected number of warnings to be generated.
+        """
```
Comment (aburrell) [inline pysatMissions/tests/test_utils.py:37]:
```suggestion
            Expected number of warnings to be generated.

        """
```

File: pysatMissions/methods/spacecraft.py (hunk 3 above)
Comment (jklenzing) [reply pysatMissions/methods/spacecraft.py:219]: This works for a time series.  I could use numpy.cross, but would have to construct 2D variables each time (as in the normalize function).

File: README.md (hunk 2 above)
Comment (jklenzing) [reply README.md:39]: I could go either way on this.  I was thinking that these two were more widely used outside of our field (similar to netCDF), so I went with common.

File: pysatMissions/methods/spacecraft.py (hunk 3 above)
Comment (jklenzing) [reply pysatMissions/methods/spacecraft.py:219]: ie, I could change the code below to use `numpy.cross`, but the function would still be needed here.

File: pysatMissions/methods/spacecraft.py [hunk 5]
```diff
@@ -173,18 +163,85 @@ def project_ecef_vector_onto_sc(inst, x_label, y_label, z_label,
...
+
+    Returns
+    -------
+    x, y, z
+        Unit vector x,y,z components
+
+    """
+    x = y1 * z2 - y2 * z1
+    y = z1 * x2 - x1 * z2
+    z = x1 * y2 - y1 * x2
```
Comment (jklenzing) [inline pysatMissions/methods/spacecraft.py:245]:
```suggestion
    vec1 = np.vstack([x1, y1, z1])
    vec2 = np.vstack([x2, y2, z2])
    x, y, z = np.cross(vec1, vec2, axis=0)
```

Comment (aburrell) [reply pysatMissions/methods/spacecraft.py:245]: I like this better.  It more clearly demonstrates why you have this helper function.

File: README.md (hunk 2 above)
Comment (aburrell) [reply README.md:39]: I was thinking more in the funding levels, not sure how well supported the maintenance is?

File: pysatMissions/methods/spacecraft.py (hunk 5 above)
Comment (jklenzing) [reply pysatMissions/methods/spacecraft.py:245]: Actually, there's probably a much nicer way of doing this at a higher level in pandas. I'm going to write this up as an issue for the next one.  For the record, the routines here are mostly copied from OMMBV to remove that dependency.

**Comment by pbhaneja** [comment]
Was able to successfully install from the test pypi server following the wiki instructions.

File: pysatMissions/methods/spacecraft.py (hunk 5 above)
Comment (jklenzing) [reply pysatMissions/methods/spacecraft.py:245]: added #96.  Experimented with fixing it now, but it's more complicated than expected.

Comment (jklenzing) [reply pysatMissions/methods/spacecraft.py]: Adding this change here.  Will improve style at a later date.

**Comment by jklenzing** [comment]
> Was able to successfully install from the test pypi server following the wiki instructions.

Thanks @pbhaneja!  Good to know this is working on windows.

File: pysatMissions/methods/spacecraft.py [hunk 6]
```diff
@@ -38,45 +37,36 @@ def add_ram_pointing_sc_attitude_vectors(inst):
...
     inst['sc_yhat_ecef_x'], inst['sc_yhat_ecef_y'], inst['sc_yhat_ecef_z'] = \
-        OMMBV.vector.cross_product(inst['sc_zhat_ecef_x'],
-                                   inst['sc_zhat_ecef_y'],
-                                   inst['sc_zhat_ecef_z'],
-                                   inst['sc_xhat_ecef_x'],
-                                   inst['sc_xhat_ecef_y'],
-                                   inst['sc_xhat_ecef_z'])
+        cross_product(inst['sc_zhat_ecef_x'], inst['sc_zhat_ecef_y'],
+                      inst['sc_zhat_ecef_z'], inst['sc_xhat_ecef_x'],
+                      inst['sc_xhat_ecef_y'], inst['sc_xhat_ecef_z'])
```
Comment (jklenzing) [inline pysatMissions/methods/spacecraft.py:57]:
```suggestion
        np.cross(inst[['sc_xhat_ecef_x', 'sc_xhat_ecef_y', 'sc_xhat_ecef_z']],
                 inst[['sc_yhat_ecef_x', 'sc_yhat_ecef_y', 'sc_yhat_ecef_z']],
                 axis=1)```

File: pysatMissions/methods/spacecraft.py [hunk 7]
```diff
@@ -38,45 +37,36 @@ def add_ram_pointing_sc_attitude_vectors(inst):
...
     inst['sc_zhat_ecef_x'], inst['sc_zhat_ecef_y'], inst['sc_zhat_ecef_z'] = \
-        OMMBV.vector.normalize(-inst['position_ecef_x'],
-                               -inst['position_ecef_y'],
-                               -inst['position_ecef_z'])
+        normalize(-inst['position_ecef_x'], -inst['position_ecef_y'],
+                  -inst['position_ecef_z'])
 
     # get y vector assuming right hand rule
     # Z x X = Y
     inst['sc_yhat_ecef_x'], inst['sc_yhat_ecef_y'], inst['sc_yhat_ecef_z'] = \
```
Comment (jklenzing) [inline pysatMissions/methods/spacecraft.py:54]:
```suggestion
    inst[['sc_zhat_ecef_x', 'sc_zhat_ecef_y', 'sc_zhat_ecef_z']] = \
```

**Review by jklenzing — COMMENTED** [review]
@aburrell, here's an alternative construction for the cross product.  This would eliminate the need for the helper function. I kind of like this better because it groups the vectors.

File: pysatMissions/methods/spacecraft.py (hunk 6 above)
Comment (jklenzing) [reply pysatMissions/methods/spacecraft.py:57]: clearly I copied to the wrong vector here, but focus on syntax.

File: README.md (hunk 2 above)
Comment (jklenzing) [reply README.md:39]: Probably a good idea to move these.  pyEphem is stale, and will be dropped in 0.4.0.

File: pysatMissions/__init__.py [hunk 8]
```diff
@@ -8,20 +8,16 @@
...
 import os
 
-try:
-    from pysatMissions import instruments
-    from pysatMissions import methods
-except ImportError as errstr:
-    logging.exception('problem importing pysatMissions: ' + str(errstr))
+from pysatMissions import instruments
+from pysatMissions import methods
+from pysatMissions import utils
```
Comment (jklenzing) [inline pysatMissions/__init__.py:18]: I don't see this in other pysat packages.  Cleaning up to standardize.

File: pysatMissions/methods/spacecraft.py [hunk 9]
```diff
@@ -173,18 +161,45 @@ def project_ecef_vector_onto_sc(inst, x_label, y_label, z_label,
 
     # TODO(#65): add checks for existence of ecef labels in inst
```
Comment (aburrell) [inline pysatMissions/methods/spacecraft.py:162]:
```suggestion
    # TODO(#65): add checks for existence of ECEF variables in the Instrument
```

**Review by rstoneback — APPROVED** [review]

**Review by aburrell — APPROVED** [review]

**Pull request merged**